scheduler/src/cook/mesos/optimizer.clj

```diff
@@ -19,122 +19,251 @@
+            :state "running"
+            :submit_time (-> job :job/submit-time .getTime)
+            :user (:job/user job)
+            :uuid (:job/uuid job)}
+           resources)))
+
+(defn create-optimizer-server
+  [{:keys [cpu-share default-runtime host-info max-waiting mem-share opt-params opt-server step-list]
+    :or {opt-params {}}
+    :as config}]
+  {:pre [(integer? default-runtime)
+         (pos? default-runtime)
+         (s/validate HostInfo host-info)
+         (integer? max-waiting)
+         (pos? max-waiting)
+         (seq step-list)]}
+  (log/info "Optimizer config" config)
+  (let [host-group 0
+        host-name->host-group (constantly host-group)
+        host-infos [host-info]
+        user->ema-mem-usage (atom {})]
+    (reify Optimizer
+      (produce-schedule [_ queued-jobs running-tasks]
+        (let [waiting-jobs (take max-waiting queued-jobs)]
+          (log/info "Optimizer inputs" {:running-tasks (count running-tasks)
+                                        :waiting-jobs (count waiting-jobs)})
+          (if (-> waiting-jobs count pos?)
+            (let [alpha 0.5
+                  now (tc/to-date (t/now))
+                  _ (log/info "Converting waiting jobs to optimizer jobs")
+                  waiting-opt-jobs (timers/time!
+                                     (timers/timer ["cook-mesos" "scheduler" "optimizer" "waiting-jobs"])
+                                     (->> waiting-jobs
+                                          (take max-waiting)
+                                          (mapv #(waiting-job->optimizer-job % default-runtime))))
+                  _ (log/info "Converting running jobs to optimizer jobs")
+                  running-opt-jobs (timers/time!
+                                     (timers/timer ["cook-mesos" "scheduler" "optimizer" "running-jobs"])
+                                     (->> running-tasks
+                                          (mapv #(running-instance->optimizer-job % default-runtime now host-name->host-group))))
+                  _ (log/info "Computing user memory usage")
+                  user->mem-usage (timers/time!
+                                    (timers/timer ["cook-mesos" "scheduler" "optimizer" "user->mem-usage"])
+                                    (->> running-opt-jobs
+                                         (group-by :user)
+                                         (pc/map-vals #(reduce + 0 (map :mem %)))
+                                         (pc/map-vals #(* (- 1 alpha) %))))
+                  _ (swap! user->ema-mem-usage
+                           (fn [ema-usage]
+                             (->> (pc/map-vals (partial * alpha) ema-usage)
+                                  (merge-with + user->mem-usage))))
+                  ;; Running must be before waiting here because optimizer determines batch order from job order
+                  opt-jobs (concat running-opt-jobs waiting-opt-jobs)
```

Comment:
Does this also imply that the intra-sequence order in `running-opt-jobs` and `waiting-opt-jobs` affects the results? Are those sorted in some meaningful way?

Reply:
Yes, their orders affect the results. Waiting jobs are sorted by the ranker. Running jobs are not sorted as we expect all running batches to fit in resource constraints and be looked into by the optimizer.

Reply:
ok